src/scripts/util.sh

```diff
@@ -37,33 +37,236 @@ error() {
+    local -n certs=$1
+    
+    for cert_name in "${!certs[@]}"; do
+        local server_names=(${certs["$cert_name"]})
+
+        local -A already_seen=()
+        local dedupped=()
+        for server_name in ${server_names[@]}; do
+            if [ -z "${already_seen[$server_name]}" ]; then
+                already_seen[$server_name]=1
+                dedupped+=($server_name)
+            fi
+        done
+
+        certs["$cert_name"]="${dedupped[@]}"
+    done
+}
+
+# A function to handle wildcard conflicts in our associative array of
+# cert_name => server_names (space-separated).
+#
+# $1: The associative bash array containing cert_name => server_names
+#     (space-separated), that will be read and updated to remove
+#     conflicting server names when one or more wildcards are present
+handle_wildcard_conflicts() {
+    local -n certs=$1
+
+    for cert_name in "${!certs[@]}"; do
+        local server_names=(${certs["$cert_name"]})
+
+        # List all the wildcards in that list, so we know which domains would be in conflict with those
+        local wildcards=()
+        for server_name in ${server_names[@]}; do
+            if [[ "$server_name" =~ ^\*\. ]]; then
+                wildcards+=("${server_name#\*.*}")
+            fi
+        done
+
+        if [ ${#wildcards[@]} -eq 0 ]; then
+            continue
+        fi
+
+        # Go over the list of domains, and keep only those that don't overlap with one of the wildcards
+        local keep_domains=()
+        for server_name in ${server_names[@]}; do
+            local conflicts_with_wildcard=0
+            for wildcard in ${wildcards[@]}; do
+                if [[ "$server_name" =~ ^[^.]*\.${wildcard}$ ]] && [[ ! "$server_name" == "*.${wildcard}" ]]; then
+                    conflicts_with_wildcard=1
+                    break
+                fi
+            done
+
+            if [ $conflicts_with_wildcard -eq 0 ]; then
+                keep_domains+=($server_name)
+            fi
+        done
+
+        certs["$cert_name"]="${keep_domains[@]}"
+    done
+}
+
+# This function forces the use of wildcards in certificates when they
+# are relevant if:
+#  (1) No `no-wildcards` pattern is found in the cert name
+# and
+#  (2.a) Either the FORCE_WILDCARDS environment variable is set to 1, or
+#  (2.b) the `force-wildcards` pattern is found in the cert name
+#
+# Wildcards will not be provisioned for domains containing less than
+# FORCE_WILDCARDS_NDOTS dots (default: 2); e.g. if FORCE_WILDCARDS_NDOTS
+# is set to 3, no automated wildcard will be created for *.example.com, but
+# some could be created for *.something.example.com.
+#
+# A wildcard will not be considered relevant if less than 2 server names
+# would be covered by it.
+#
+# $1: The associative bash array containing cert_name => server_names
+#     (space-separated), that will be read and updated to remove
+#     conflicting server names when one or more wildcards are present
+force_wildcards() {
```

Comment:
So that pretty much handles what's exactly described in #92 - And allows to enable/disable it globally, or per-certificate :)